c-dependencies/js-compute-runtime/builtins/backend.cpp

```diff
@@ -0,0 +1,705 @@
+#include <iostream>
+#include <optional>
+
+// TODO: remove these once the warnings are fixed
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winvalid-offsetof"
+#include "js/experimental/TypedData.h"
+#pragma clang diagnostic pop
+
+#include "backend.h"
+#include "host_call.h"
+#include "js/Conversions.h"
+
+bool isDigit(char character) { return character >= 48 && character <= 57; }
+bool isLowerAlpha(char character) { return character >= 97 && character <= 122; }
+bool isUpperAlpha(char character) { return character >= 65 && character <= 90; }
+bool isAlphaNumeric(char character) {
+  if (!isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character)) {
+    return false;
+  }
+  return true;
+}
+bool isDash(char character) { return character == 45; }
+bool isDot(char character) { return character == 46; }
+
+bool isNotAlphaNumericOrDash(char character) {
+  return !isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character) &&
+         !isDash(character);
+}
+
+bool isNotAlphaNumericDotOrDash(char character) {
+  return !isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character) &&
+         !isDash(character) && !isDot(character);
+}
+
+bool isValidHost(std::string host) {
+  // ValidHostRegex =
+  // "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])(:[0-9]+)$";
+  auto firstCharacter = host.front();
+  // check first character is in the regex [a-zA-Z0-9]
+  if (!isAlphaNumeric(firstCharacter)) {
+    return false;
+  }
+  // split the hostname from the port
+  int pos = host.find_first_of(':');
+  std::string hostname = host.substr(0, pos);
+
+  auto lastCharacter = hostname.back();
+  // check last character is in the regex [a-zA-Z0-9]
+  if (!isAlphaNumeric(lastCharacter)) {
+    return false;
+  }
+
+  // find the position of the last .
+  auto lastDot = hostname.rfind('.');
+  if (lastDot == std::string::npos) {
+    return true;
+  }
+
+  // check the character before the last . is in the regex [a-zA-Z0-9]
+  if (!isAlphaNumeric(hostname.at(lastDot - 1))) {
+    return false;
+  }
+  // check all other characters before the last . are in the regex [a-zA-Z0-9\-.]
+  auto last = std::next(hostname.begin(), lastDot);
+  auto it = std::find_if(std::next(hostname.begin()), last, isNotAlphaNumericDotOrDash);
+
+  if (it != last) {
+    return false;
+  }
+  // check all characters after the last . (but not the last character) are in the regex
+  // [a-zA-Z0-9\-]
+  last = std::prev(hostname.end());
+  it = std::find_if(std::next(hostname.begin(), lastDot + 1), last, isNotAlphaNumericOrDash);
+  if (it != last) {
+    return false;
+  }
+  // if there is a port - confirm it is all digits and is between 0 and 65536
+  if (pos != std::string::npos) {
+    std::string port = host.substr(pos + 1);
```

Comment:
This will probably allocate a string, you could avoid it by using a `std::string_view`.

Reply:
you spotted this one as well! 😄 This has changed to std::string_view in [`5fcb9df` (#250)](https://github.com/fastly/js-compute-runtime/pull/250/commits/5fcb9dfdfa86026529ea933260a5c8936d9e922f)